Site/dataobjects/SiteAttachment.php

```diff
@@ -1,556 +1,448 @@
+
+    public function setFileBase($file_base)
+    {
+        $this->file_base = $file_base;
+    }
+
+    public function getFileDirectory()
+    {
+        $items = [$this->getFileBase(), $this->getAttachmentSet()->shortname];
+
+        return implode(DIRECTORY_SEPARATOR, $items);
+    }
+
+    public function getFilePath()
+    {
+        $items = [$this->getFileDirectory(), $this->getFilename()];
+
+        return implode(DIRECTORY_SEPARATOR, $items);
+    }
+
+    public function getFilename()
+    {
+        if ($this->getAttachmentSet()->obfuscate_filename) {
+            $prefix = $this->obfuscated_id;
+        } else {
+            $prefix = $this->id;
+        }
+
+        return sprintf('%s.%s', $prefix, $this->getExtension());
+    }
+
+    public function getHttpHeaders()
+    {
+        $headers = [];
+
+        // Set a "never-expire" policy with a far future max age (10 years) as
+        // suggested https://developer.yahoo.com/performance/rules.html#expires.
+        // As well, set Cache-Control to public, as this allows some browsers to
+        // cache the images to disk while on https, which is a good win. This
+        // depends on setting new object ids when updating the object, if this
+        // isn't true of a subclass this will have to be overwritten.
+        $headers['Cache-Control'] = 'public, max-age=315360000';
+
+        $headers['Content-Type'] = $this->mime_type;
+        $headers['Content-Length'] = $this->file_size;
+
+        // Convert to an ASCII string. Approximate non ACSII characters.
+        $filename = iconv(
+            'UTF-8',
+            'ASCII//TRANSLIT',
+            ($this->human_filename != '') ?
+                $this->human_filename :
+                $this->original_filename
+        );
+
+        // Format the filename according to the qtext syntax in RFC 822
+        $filename = str_replace(
+            ['\\', "\r", '"'],
+            ['\\\\', "\\\r", '\\"'],
+            $filename
+        );
+
+        $headers['Content-Disposition'] = sprintf(
+            'attachment; filename="%s"',
+            $filename
+        );
+
+        return $headers;
+    }
+
+    public function load($id)
+    {
+        $loaded = parent::load($id);
+
+        if ($loaded && $this->attachment_set_shortname !== null) {
+            if ($this->attachment_set->shortname !==
+                $this->attachment_set_shortname) {
+                throw new SiteException('Trying to load attachment with the ' .
+                    'wrong attachment set. This may happen if the wrong ' .
+                    'wrapper class is used.');
+            }
+        }
+
+        return $loaded;
+    }
+
+    public function process($file_path)
+    {
+        $this->checkDB();
+
+        try {
+            $transaction = new SwatDBTransaction($this->db);
+
+            if ($this->getAttachmentSet()->obfuscate_filename) {
+                $this->obfuscated_id = sha1(uniqid(random_int(0, mt_getrandmax()), true));
```

Comment:
Arguable should be extracted to a separate static class function. This is fine for now but I think we might do similar generation in a few places a few different ways.

Reply:
Refactored to `Site::generateRandomHash()` method.